Avoid automatic saving of uncontrolled Nav blocks

getdave opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR updates the Nav block so that if it has uncontrolled blocks (i.e. those not coming from a `wp_navigation` entity but rather those provided directly as `<-- wp:block`) then it will _not_ automatically convert them into an entity.

Instead the block will now wait for the blocks to become "dirty" (i.e. have been changed) before converting into a `wp_navigation`.

Part of https://github.com/WordPress/gutenberg/issues/38157 and/or https://github.com/WordPress/gutenberg/issues/39489

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

There are a number of ways where a Nav block might contain raw inner blocks. For example:
- a Theme's template part
- a Pattern

Taking the former example [from TT2 here is the block definition](https://github.com/WordPress/twentytwentytwo/blob/27054dbcb95f5d72cd943a0f3318c3ad97599f85/parts/header.html#L9-L11):

```
<!-- wp:navigation {"layout":{"type":"flex","setCascadingProperties":true,"justifyContent":"right"}} -->
<!-- wp:page-list {"isNavigationChild":true,"showSubmenuIcon":true,"openSubmenusOnClick":false} /-->
<!-- /wp:navigation -->
```

Notice how the Page List block is provided in order that there is default content in the block when the Site Editor is loaded.

The current problem with this is that as soon as the user interacts with the Nav block it will automatically convert the uncontrolled blocks into a `wp_navigation` post. There is no warning that this has occured and so the user is none the wiser.

The result of this is that it's super easy to inadvertently create many `wp_navigation` posts simply by selecting Navigation blocks on your site.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

This PR fixes thing by allowing the raw inner blocks to stay uncontrolled until such time as the user makes a _modification_ to those blocks. As soon as that occurs then they will be auto-converted into a `wp_navigation` post and a notice will be displayed announcing that to the user.

In future it might be nice to give the user some means to opt in/out of the conversion perhaps via a dialog. This PR follows te simplest path to fix the core problem of accidental creation of `wp_navigation` posts.

## Testing Instructions

- Create New Post
- Switch to Code View
- Paste
```
<!-- wp:navigation {"layout":{"type":"flex","setCascadingProperties":true,"justifyContent":"right"}} -->
<!-- wp:page-list {"isNavigationChild":true,"showSubmenuIcon":true,"openSubmenusOnClick":false} /-->
<!-- /wp:navigation -->
```
- Switch back to visual mode.
- See Nav block with Page List block in canvas. 
- Select Nav block (this would previously have triggered conversion to an entity).
- Switch back to Code view and check inner blocks - they should be preserved. There should be no `ref` attribute on the Nav block.
- Switch back to Visual Mode.
- Click on Page List block. Click `Edit` and then `Convert`.
- This will trigger a "change" to the inner blocks.
- See loading state - an `wp_navigation` entity will be created.
- Switch to Code view.
- See inner blocks are gone and there is now a `ref` attribute on the Nav block.

## Screenshots or screencast <!-- if applicable -->

getdave committed:
Only convert uncontrolled inner blocks if there are changes

getdave committed:
Show notification on conversion to entity

getdave committed:
Disable inner blocks only when being converted

getdave committed:
Limit blocks to match controlled inner blocks

getdave committed:
Remove Disable state on uncontrolled inner blocks

getdave commented:
@talldan I'm particularly pinging you here because I know you did a lot of work on handling uncontrolled inner blocks and you're very likely to be able to notice edge cases 🙏

getdave commented:
...I need to fix the e2e tests and add a new one to assert that making inner blocks dirty will trigger conversion to entity.

scruffian commented:
I'm seeing a navigation menu being created when I select a block in a template.

1. Delete all navigation menus
2. Load Archeo in the Site Editor
3. Select the page list block
4. Notice that a new navigation is created:

https://user-images.githubusercontent.com/275961/161268525-68f5e393-e9bf-429c-88d9-e350044e7bdf.mp4

getdave commented:
> I'm seeing a navigation menu being created when I select a block in a template.

@scruffian I re-tested this and I couldn't replicate what you saw. Are you able to re-test

https://user-images.githubusercontent.com/444434/161293853-1200a184-61a0-464d-b342-81093485eeb2.mov

If the problem persists then would you be open to pairing on testing this early next week?

getdave commented on packages/block-library/src/navigation/edit/unsaved-inner-blocks.js:
```diff
@@ -33,13 +49,41 @@ export default function UnsavedInnerBlocks( {
+		originalBlocks.current = blocks;
+	}
+
+	// If the current inner blocks object is different in any way
+	// from the original inner blocks from the post content then the
+	// user has made changes to the inner blocks. At this point the inner
```
I should update this comment. It's not about whether object contents are different. It's about the object references changing if the block data changes.

scruffian commented:
> @scruffian I re-tested this and I couldn't replicate what you saw. Are you able to re-test

I also can't replicate now sorry - there must have been a problem in my setup...

scruffian commented on packages/block-library/src/navigation/edit/unsaved-inner-blocks.js:
```diff
@@ -33,13 +49,41 @@ export default function UnsavedInnerBlocks( {
+
+	// Initially store the uncontrolled inner blocks for
+	// dirty state comparison.
+	if ( ! originalBlocks?.current ) {
```
Is modifying the ref directly like this advised? Would it be better to put the ref directly on the component?

getdave replied:
What makes you concerned here?

The ref is a stored value which we create solely to store whatever blocks the component is passed as its initial value.

We do nothing with `originalBlocks` in terms of passing it to `<InnerBlocks>`. Its sole purpose is to act as a source of truth for the original "state" of the blocks that was originally passed to the component. 

I'm going to double check this checks out correctly and I'm not setting us up for bugs.

scruffian replied:
I guess my concern is whether we are abusing the `ref` feature of React - why not just us a local variable?

getdave replied:
Because a ref is guaranteed to be the same reference for the _lifetime_ of the component whereas the local variable will be a new reference on each render of the component:

> [The returned object will persist for the full lifetime of the component](https://reactjs.org/docs/hooks-reference.html#useref) (React Docs)

getdave replied:
So basically I'm using the ref to store the "source of truth" for the lifecycle of the component. Then if at any point the current blocks don't match up with the original source of truth reference stored in the `ref` then the blocks have changed because the references are different.

adamziel replied:
Would it be possible to replace this ref with a combination of `hasUncontrolledInnerBlocks` and `hasUnsavedBlocks` that's used in edit/index.js?

adamziel replied:
Basically I wonder if this information can be derived from somewhere instead of computing it here

getdave replied:
`hasUnsavedBlocks` is computed here as a product of
- whether there is an entity 
- whether there are uncontrolled inner blocks present (this is determined by `getBlock( clientId ).innerBlocks`)

https://github.com/WordPress/gutenberg/blob/830c4ed8f82396980a8d415ea3d2f56150a68f16/packages/block-library/src/navigation/edit/index.js#L544

Neither of those factors tell us whether the inner blocks have had _modifications_. They simply report that

- there is no `wp_navigation` entity associated with the current block.
- there are inner blocks present.

But can't we use the "saved" state to determine whether there are changes to the blocks? Unfortunately not. Why? Because that's [putting the cart before the horse](https://idioms.thefreedictionary.com/put+the+cart+before+the+horse). 

We need the uncontrolled version of the Nav block to save (convert to entity mode) **only when we detect there have been modifications** to those blocks. Therefore we can't use the "saved" state to determine whether the blocks have changed because that state should be a product of whether the blocks have changed.

As far as I can see the only way to determine if the uncontrolled inner blocks have changed is to effectively memoize the _initial_ value of blocks (i.e. `getBlock( clientId ).innerBlocks`) and then check that to see if it changes.

This is what I'm trying to do with the ref. I save the initial value and then compare that reference against the new incoming value. When the user makes a change to the inner blocks the object returned by `getBlock( clientId ).innerBlocks` will change and thus we can compare the references and determine they are no longer equal.

I might be able to do that outside the component but I'm not sure what the value of that would be.

adamziel replied:
Wow, great exploration @getdave 👍  alright, let's stick to using the ref then. If this happens more often, then we could look into abstracting it out into a reusable API.

scruffian commented on packages/block-library/src/navigation/edit/unsaved-inner-blocks.js:
```diff
@@ -33,13 +49,41 @@ export default function UnsavedInnerBlocks( {
+
+	const shouldDirectInsert = useMemo(
+		() =>
+			blocks.every(
```
TIL `every`

scruffian commented on packages/block-library/src/navigation/edit/unsaved-inner-blocks.js:
```diff
@@ -33,13 +49,41 @@ export default function UnsavedInnerBlocks( {
 
 	const innerBlocksProps = useInnerBlocksProps( blockProps, {
-		renderAppender: hasSelection ? undefined : false,
+		// renderAppender: hasSelection ? undefined : false,
```
Should we remove this line?

getdave replied:
I think we need it. Reinstated.

scruffian commented on packages/block-library/src/navigation/edit/unsaved-inner-blocks.js:
```diff
@@ -120,17 +165,17 @@ export default function UnsavedInnerBlocks( {
 	return (
 		<div className="wp-block-navigation__unsaved-changes">
-			<Disabled
+			<div
```
Do we still want this to be disabled in some case?

getdave replied:
Yeh I was thinking if `isSaving` is true then we want it to be disabled I guess.

getdave replied:
Updated so it renders a div or `Disabled` component if saving.

getdave committed:
Avoiding setting ref in component body

React docs explain this can lead to undesired side effects

getdave committed:
Match render appender with normal inner blocks

getdave committed:
Disable inner blocks when saving

getdave committed:
Add test

talldan commented on packages/block-library/src/navigation/edit/unsaved-inner-blocks.js:
```diff
@@ -118,19 +171,23 @@ export default function UnsavedInnerBlocks( {
-	return (
-		<div className="wp-block-navigation__unsaved-changes">
-			<Disabled
+	return createElement(
+		isSaving ? Disabled : 'div', // when saving make component disabled.
+		{
+			className: 'wp-block-navigation__unsaved-changes',
+		},
```
Would be good to avoid mixing `createElement` and JSX. I think something like this works:
```jsx
const Wrapper = isSaving ? Disabled : 'div';

return (
    <Wrapper className="wp-block-navigation__unsaved-changes">
        // ...
    </Wrapper>
);

getdave replied:
I did try that but I was getting an error about `div`. Let me double check I wasn't making a typo late yesterday...

getdave replied:
Thanks Dan. That worked.

Annoyingly this is the second PR where I think code `watch` mode has failed me and lead me to think that something isn't working when in fact it is.

🙇

getdave committed:
Avoid createElement

Kudos @talldan https://github.com/WordPress/gutenberg/pull/39883#discussion_r842271516

getdave committed:
Update tests to check modifying inner blocks creates entity

getdave committed:
Remove invalid test

You cannot reset a block that has uncontrolled inner blocks

getdave committed:
Improve comment for purpose of uncontrolled blocks

getdave committed:
Check for button being enable before converting Page List

getdave committed:
Update snapshots

scruffian reviewed (approved):
This looks like a good improvement. Ideally the new navigation shouldn't be created until I hit the save button. I added an issue for that here: https://github.com/WordPress/gutenberg/issues/40050

draganescu commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -536,11 +536,12 @@ function Navigation( {
+	// If the block has inner blocks, but no menu id, then these blocks are either:
+	// - inserted via a pattern.
+	// - inserted directly via Code View (or otherwise).
+	// - from an older version of navigation block added before the block used a wp_navigation entity.
```
Can't they be added directly by the theme as well?

draganescu commented:
This makes adding a new item quite weird:

https://user-images.githubusercontent.com/107534/161789641-bfab0676-986e-4195-ae7b-d4410edc983e.mp4
